cluster-app-values-schema/README.md

```diff
@@ -0,0 +1,384 @@
+- [R9: Required properties must be marked as such](#r9)
+- [R10: Use `anyOf` and `oneOf` only for specific purposes](#r10)
+- [R11: Use `deprecated` to phase out properties](#r11)
+- [R12: Provide valid string values where possible](#r12)
+- [R13: Avoid recursion](#r13)
+- [R14: Avoid logical constructs using `if`, `then`, `else`](#r14)
+- [R15: Avoid `unevaluatedProperties`, `unevaluatedItems`](#r15)
+- [R16: Array items and tuple validation](#r16)
+- [R17: Common schema structure](#r17)
+
+## Background
+
+TODO: Some more info regarding the rationale.
+
+- In-place documentation of properties
+- Better validation to catch configuration errors before they are applied
+- Generated entry forms
+- Generated information display
+- Enabling processes and communication about explicit contracts instead of implicit behaviour
+
+## Definitions
+
+- **Cluster app**: App (as defined by the Giant Swarm app platform) that allows for provisioning a cluster in a Giant Swarm installation.
+- **JSON Schema**: declarative language that allows to annotate and validate documents that are compatible with JSON.
+- The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html)
+
+## Requirements
+
+Requirements carry identifiers with prefix `R` and a unique number, for easy referencing.
+
+### R1: JSON Schema dialect (draft 2020-12) must be specified {#r1}
+
+Each cluster app's values schema file MUST specify the schema dialect (also called draft) used via the `$schema` keyword on the top level, with the draft URI as a value.
+
+The draft URI MUST be `https://json-schema.org/draft/2020-12/schema`.
+
+Example:
+
+```json
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  ...
+}
+```
+
+### R2: A single type must be declared {#r2}
+
+For each property, including the root level schema, the `type` keyword MUST be present, and there MUST be a single value for the `type` keyword.
+
+While JSON Schema draft 2020-12 allows multiple values for the `type` keyword, for cluster-apps this would complicate the generation of user interfaces. Hence we require a single type to be specified.
+
+For validation, there is no difference between an array with a single member or a single string. The two following examples can be considered identical.
+
+```json
+{
+  "type": "string",
+  "title": "Name"
+}
+```
+
+```json
+{
+  "type": ["string"],
+  "title": "Name"
+}
+```
+
+### R3: Explicitly cover all allowed properties {#r3}
+
+All properties that can be used in values MUST be covered in the schema explicitly.
+
+To enforce this and to disable the use of any undefined properties, the keyword `additionalProperties` MUST be set to `false` on the schema's root object.
+
+For objects deeper in the schema hierarchy, keyword `additionalProperties` SHOULD be set to `false`, too.
```

Comment:
I think here we need to adapt. `additionalProperties` can also be set to a schema.
It should be always set and never to `true`.

Reply:
Okay, that was a beginner question I guess 😆 – `additionalProperties` is for when the object keys can be dynamic, such as in `nodePools: { abc: ..., def: ... }`.

Reply:
Exactly. `additionalProperties` can be used if the keys can be arbitrary and there is also `patternProperties`, which uses a regex pattern to restrict the keys. This one is also used in our schemas, AFAIK.